libraries/platforms/dromajo-vcs/software/src/bp_utils.c

```diff
@@ -0,0 +1,120 @@
+// BlackParrot utilities' definitions
+
+#include <stdint.h>
+#include <bp_utils.h>
+
+/*
+ * Get hardware thread ID
+ * @returns result of reading the mhartid register
+ */
+uint64_t bp_get_hart() {
+    uint64_t core_id;
+    __asm__ volatile("csrr %0, mhartid": "=r"(core_id): :);
+    return core_id;
+}
+
+/*
+ * Checks to see if a barrier is finished
+ * @param[in] barrier_address --> An address in memory that all cores to write to after hitting the barrier
+ * @param[in] total_num_cores --> Number of cores
+ */
+void bp_barrier_end(volatile uint64_t * barrier_address, uint64_t total_num_cores) {
+    uint64_t core_id;
+    uint64_t atomic_inc = 1;
+    uint64_t atomic_result;
+    __asm__ volatile("csrr %0, mhartid": "=r"(core_id): :);
+    
+    /* if we're not core 0, increment the barrier and then just loop */
+    if (core_id != 0) {
+        __asm__ volatile("amoadd.d %0, %2, (%1)": "=r"(atomic_result) 
+                                                : "r"(barrier_address), "r"(atomic_inc)
+                                                :);
+        while (1) { }
+    }
+    /* 
+     * if we're core 0, increment the barrier as well and then test if the
+     * barrier is equal to the total number of cores
+     */
+    else {
+        uint64_t finish_value = 0;
+        __asm__ volatile("amoadd.d %0, %2, (%1)": "=r"(atomic_result) 
+                                                : "r"(barrier_address), "r"(atomic_inc)
+                                                :);
+        while(*barrier_address < total_num_cores) {
+
+            
+        }
+        bp_finish(0);
+    }
+}
+
+/*
+ * Sends a finish packet to the host
+ * @param[in] code --> Finish code
+ */
+void bp_finish(int16_t code) {
+  uint64_t core_id;
+
+	// Fixme: Core IDs for all BP cores might be zero. How do we handle this?
+  __asm__ volatile("csrr %0, mhartid": "=r"(core_id): :);
+
+  // Finish packet contains core id at the top 16 bits and
+	// finish code at the bottom 16 bits
+	// zero --> success code
+	// nonzero --> fail code
+	hb_mc_packet_t finish_pkt;
+	finish_pkt.request.x_dst = HOST_X_COORD;
+	finish_pkt.request.y_dst = HOST_Y_COORD;
+	finish_pkt.request.x_src = BP_HOST_LINK_X;
+	finish_pkt.request.y_src = BP_HOST_LINK_Y;
+	finish_pkt.request.op_v2 = HB_MC_PACKET_OP_REMOTE_SW;
+	finish_pkt.request.payload = (core_id << 16) | (0x0000FFFF & code);
```

Comment:
Replace magic numbers

Reply:
I am attempting to encode the core ID using the top 16 bits (currently everything is 0 since there is just 1 core but might be useful later) and the finish code in the bottom 16 bits. I have mentioned this in the comments above. I don't think this requires a macro.

Reply:
I think everything requires a macro

Reply:
But this can be left as is